### ISSUE TITLE
App crashes on many devices (All on Android 4.0 - 4.4)

### ISSUE THREAD

**Issue by liutingdu** [evercam/evercam-play-android#89]

**Comment by liutingdu** [comment]
Reproduced it on Simulator, it crashes with third party libraries: SplunkMint, Mixpanel and Unirest. But after upgrading all those libraries, got new error: 
```
08-04 15:18:03.225    4473-4473/io.evercam.androidapp E/AndroidRuntime﹕ FATAL EXCEPTION: main
    java.lang.RuntimeException: Unable to start activity ComponentInfo{io.evercam.androidapp/io.evercam.androidapp.CamerasActivity}: android.view.InflateException: Binary XML file line #14: Error inflating class io.evercam.androidapp.custom.CustomScrollView
            at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2059)
            at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2084)
            at android.app.ActivityThread.access$600(ActivityThread.java:130)
            at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1195)
            at android.os.Handler.dispatchMessage(Handler.java:99)
            at android.os.Looper.loop(Looper.java:137)
            at android.app.ActivityThread.main(ActivityThread.java:4745)
            at java.lang.reflect.Method.invokeNative(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:511)
            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:786)
            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:553)
            at dalvik.system.NativeStart.main(Native Method)
     Caused by: android.view.InflateException: Binary XML file line #14: Error inflating class io.evercam.androidapp.custom.CustomScrollView
            at android.view.LayoutInflater.createView(LayoutInflater.java:613)
            at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:687)
            at android.view.LayoutInflater.rInflate(LayoutInflater.java:746)
            at android.view.LayoutInflater.inflate(LayoutInflater.java:489)
            at android.view.LayoutInflater.inflate(LayoutInflater.java:396)
            at android.view.LayoutInflater.inflate(LayoutInflater.java:352)
            at com.android.internal.policy.impl.PhoneWindow.setContentView(PhoneWindow.java:256)
            at android.app.Activity.setContentView(Activity.java:1867)
            at io.evercam.androidapp.CamerasActivity.onCreate(CamerasActivity.java:91)
            at android.app.Activity.performCreate(Activity.java:5008)
            at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1079)
            at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2023)
            at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2084)
            at android.app.ActivityThread.access$600(ActivityThread.java:130)
            at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1195)
            at android.os.Handler.dispatchMessage(Handler.java:99)
            at android.os.Looper.loop(Looper.java:137)
            at android.app.ActivityThread.main(ActivityThread.java:4745)
            at java.lang.reflect.Method.invokeNative(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:511)
            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:786)
            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:553)
            at dalvik.system.NativeStart.main(Native Method)
     Caused by: java.lang.reflect.InvocationTargetException
            at java.lang.reflect.Constructor.constructNative(Native Method)
            at java.lang.reflect.Constructor.newInstance(Constructor.java:417)
            at android.view.LayoutInflater.createView(LayoutInflater.java:587)
            at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:687)
            at android.view.LayoutInflater.rInflate(LayoutInflater.java:746)
            at android.view.LayoutInflater.inflate(LayoutInflater.java:489)
            at android.view.LayoutInflater.inflate(LayoutInflater.java:396)
            at android.view.LayoutInflater.inflate(LayoutInflater.java:352)
            at com.android.internal.policy.impl.PhoneWindow.setContentView(PhoneWindow.java:256)
            at android.app.Activity.setContentView(Activity.java:1867)
            at io.evercam.androidapp.CamerasActivity.onCreate(CamerasActivity.java:91)
            at android.app.Activity.performCreate(Activity.java:5008)
            at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1079)
            at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2023)
            at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2084)
            at android.app.ActivityThread.access$600(ActivityThread.java:130)
            at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1195)
            at android.os.Handler.dispatchMessage(Handler.java:99)
            at android.os.Looper.loop(Looper.java:137)
            at android.app.ActivityThread.main(ActivityThread.java:4745)
            at java.lang.reflect.Method.invokeNative(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:511)
            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:786)
            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:553)
            at dalvik.system.NativeStart.main(Native Method)
     Caused by: java.lang.NoClassDefFoundError: io.evercam.androidapp.custom.CustomScrollView$1
            at io.evercam.androidapp.custom.CustomScrollView.<init>(CustomScrollView.java:26)
            at java.lang.reflect.Constructor.constructNative(Native Method)
            at java.lang.reflect.Constructor.newInstance(Constructor.java:417)
            at android.view.LayoutInflater.createView(LayoutInflater.java:587)
            at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java:687)
            at android.view.LayoutInflater.rInflate(LayoutInflater.java:746)
            at android.view.LayoutInflater.inflate(LayoutInflater.java:489)
            at android.view.LayoutInflater.inflate(LayoutInflater.java:396)
            at android.view.LayoutInflater.inflate(LayoutInflater.java:352)
            at com.android.internal.policy.impl.PhoneWindow.setContentView(PhoneWindow.java:256)
            at android.app.Activity.setContentView(Activity.java:1867)
            at io.evercam.androidapp.CamerasActivity.onCreate(CamerasActivity.java:91)
            at android.app.Activity.performCreate(Activity.java:5008)
            at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1079)
            at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2023)
            at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2084)
            at android.app.ActivityThread.access$600(ActivityThread.java:130)
            at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1195)
            at android.os.Handler.dispatchMessage(Handler.java:99)
            at android.os.Looper.loop(Looper.java:137)
            at android.app.ActivityThread.main(ActivityThread.java:4745)
            at java.lang.reflect.Method.invokeNative(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:511)
            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:786)
            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:553)
            at dalvik.system.NativeStart.main(Native Method)
```

It needs to be fixed.

**Comment by liutingdu** [comment]
It seems crashes when creating a runnable in the CustomScrollView, don't understand why but can be temporarily fixed by removing the runnable (it's not in use at this moment).

Then getting new crash with the following logs: 
```
08-05 05:28:51.071    3080-3080/? E/dalvikvm﹕ Could not find class 'io.evercam.androidapp.custom.CameraLayout$3', referenced from method io.evercam.androidapp.custom.CameraLayout.<init>
    08-05 05:28:51.071    3080-3080/? W/dalvikvm﹕ VFY: unable to resolve new-instance 6631 (Lio/evercam/androidapp/custom/CameraLayout$3;) in Lio/evercam/androidapp/custom/CameraLayout;
    08-05 05:28:51.071    3080-3080/? D/dalvikvm﹕ VFY: replacing opcode 0x22 at 0x0019
    08-05 05:28:51.071    3080-3080/? W/dalvikvm﹕ VFY: unable to find class referenced in signature (Lio/evercam/androidapp/custom/GradientTitleLayout;)
    08-05 05:28:51.071    3080-3080/? E/dalvikvm﹕ Could not find class 'io.evercam.androidapp.custom.CameraLayout$2', referenced from method io.evercam.androidapp.custom.CameraLayout.showOfflineIcon
    08-05 05:28:51.071    3080-3080/? W/dalvikvm﹕ VFY: unable to resolve new-instance 6630 (Lio/evercam/androidapp/custom/CameraLayout$2;) in Lio/evercam/androidapp/custom/CameraLayout;
    08-05 05:28:51.071    3080-3080/? D/dalvikvm﹕ VFY: replacing opcode 0x22 at 0x0005
    08-05 05:28:51.071    3080-3080/? I/dalvikvm﹕ Could not find method io.evercam.androidapp.custom.GradientTitleLayout.getOfflineImageView, referenced from method io.evercam.androidapp.custom.CameraLayout.getOfflineIconBounds
    08-05 05:28:51.071    3080-3080/? W/dalvikvm﹕ VFY: unable to resolve virtual method 44814: Lio/evercam/androidapp/custom/GradientTitleLayout;.getOfflineImageView ()Landroid/widget/ImageView;
    08-05 05:28:51.071    3080-3080/? D/dalvikvm﹕ VFY: replacing opcode 0x6e at 0x0007
    08-05 05:28:51.075    3080-3080/? I/dalvikvm﹕ Could not find method io.evercam.androidapp.custom.GradientTitleLayout.removeGradientShadow, referenced from method io.evercam.androidapp.custom.CameraLayout.showThumbnail
    08-05 05:28:51.075    3080-3080/? W/dalvikvm﹕ VFY: unable to resolve virtual method 44815: Lio/evercam/androidapp/custom/GradientTitleLayout;.removeGradientShadow ()V
    08-05 05:28:51.075    3080-3080/? D/dalvikvm﹕ VFY: replacing opcode 0x6e at 0x0057
    08-05 05:28:51.075    3080-3080/? I/dalvikvm﹕ Could not find method io.evercam.androidapp.custom.GradientTitleLayout.setTitle, referenced from method io.evercam.androidapp.custom.CameraLayout.updateTitleIfDifferent
    08-05 05:28:51.075    3080-3080/? W/dalvikvm﹕ VFY: unable to resolve virtual method 44816: Lio/evercam/androidapp/custom/GradientTitleLayout;.setTitle (Ljava/lang/String;)V
    08-05 05:28:51.075    3080-3080/? D/dalvikvm﹕ VFY: replacing opcode 0x6e at 0x0028
    08-05 05:28:51.075    3080-3080/? D/dalvikvm﹕ DexOpt: unable to opt direct call 0xaeaa at 0x1b in Lio/evercam/androidapp/custom/CameraLayout;.<init>
    08-05 05:28:51.075    3080-3080/? D/dalvikvm﹕ DexOpt: unable to opt direct call 0xaf0d at 0xdf in Lio/evercam/androidapp/custom/CameraLayout;.<init>
    08-05 05:28:51.075    3080-3080/? D/dalvikvm﹕ DexOpt: unable to opt direct call 0xaea8 at 0x105 in Lio/evercam/androidapp/custom/CameraLayout;.<init>
    08-05 05:28:51.075    3080-3080/? D/dalvikvm﹕ DexOpt: unable to opt direct call 0xaea9 at 0x07 in Lio/evercam/androidapp/custom/CameraLayout;.showOfflineIcon
    08-05 05:28:51.075    3080-3080/? D/AndroidRuntime﹕ Shutting down VM
    08-05 05:28:51.075    3080-3080/? W/dalvikvm﹕ threadid=1: thread exiting with uncaught exception (group=0xa4b68648)
    08-05 05:28:52.767    3080-3104/io.evercam.androidapp D/dalvikvm﹕ GC_FOR_ALLOC freed 778K, 7% free 12359K/13164K, paused 7ms, total 7ms
    08-05 05:28:54.083    3080-3080/io.evercam.androidapp E/AndroidRuntime﹕ FATAL EXCEPTION: main
        java.lang.NoClassDefFoundError: io.evercam.androidapp.custom.CameraLayout$3
                at io.evercam.androidapp.custom.CameraLayout.<init>(CameraLayout.java:256)
                at io.evercam.androidapp.CamerasActivity.addAllCameraViews(CamerasActivity.java:439)
                at io.evercam.androidapp.CamerasActivity$1.run(CamerasActivity.java:123)
                at android.os.Handler.handleCallback(Handler.java:730)
                at android.os.Handler.dispatchMessage(Handler.java:92)
                at android.os.Looper.loop(Looper.java:137)
                at android.app.ActivityThread.main(ActivityThread.java:5103)
                at java.lang.reflect.Method.invokeNative(Native Method)
                at java.lang.reflect.Method.invoke(Method.java:525)
                at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:737)
                at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:553)
                at dalvik.system.NativeStart.main(Native Method)
    08-05 05:31:12.083    3080-3080/? I/Process﹕ Sending signal. PID: 3080 SIG: 9
```

Posted question here: http://stackoverflow.com/questions/31829350/app-crashes-with-error-could-not-find-class-xxxx-referenced-from-method-xxx

**Comment by liutingdu** [comment]
It turns out to be a multidex problem, fixed in 1.4.9
